### ISSUE TITLE
manubot cite json output order does not match order of input identifiers

### ISSUE THREAD

**Issue by vincerubinetti** [manubot/manubot#240]
On Windows, I'm running the following command:
```
manubot cite --output research-2015.json \
  doi:10.1093/hmg/ddv440 \
  doi:10.1186/s12864-015-2250-5 \
  doi:10.1038/ncomms9019 \
  doi:10.1186/s13040-015-0050-8 \
  doi:10.1038/ng.3259 \
  doi:10.1038/nmeth.3249 \
  doi:10.1371/journal.pcbi.1004005 \
  pmc:PMC4299935 \
```
and the order of the json data that comes back is this:
```
10.1038/ncomms9019
10.1038/ng.3259
10.1038/nmeth.3249 
10.1093/hmg/ddv440
10.1186/s12864-015-2250-5
10.1186/s13040-015-0050-8
10.1371/journal.pcbi.1004005
PMC4299935
```
It looks like the results are returned in alphabetical order of their identifier? But perhaps it should match the order they are provided in. This would make programming with the command easier, avoiding additional step of matching the output back to the order of the input.

**Comment by agitter** [comment]
It looks like this happens because `Citations.get_csl_items` calls `group_citekeys_by`:
https://github.com/manubot/manubot/blob/a57ccf0be6972329ff3010eaaa0c5df7ccebb2d5/manubot/cite/citations.py#L56-L68

The `sorted` call is there because `groupby` expects the iterable to already be sorted on the key.  To maintain the original order, I think we may have to explicitly store the original order and resort after `groupby`.

**Comment by vincerubinetti** [comment]
Not a huge issue, but would make life easier in certain cases. At the very least, it should probably be documented in the readme.

Here's my use case, fwiw:

I'm redoing the [greenelab website](https://greenelab.github.io/greenelab.com/). For the list of publications we had, maintainers were previously just manually formatting and pasting in the citations. In the new version, we're using `manubot cite` to automatically get all the meta data like title, author, etc.

However, we want to sort the publications by date, and the templating language we're using to display the `manubot cite` json output is not good at sorting by deeply nested fields in a json object, like `issued: { "date-parts: [ [2020,11, 23] ] }`. So we were relying on the order we specified the identifiers.

**Comment by dhimmel** [comment]
Yeah, the old implementation returned the references in the same order as the input. Let me look into the easiest way to restore the old behavior.

A related question, if multiple citations are provided to the same work, should they be consolidated? Example: `manubot cite doi:10.1093/hmg/ddv440 doi:10.1093/HMG/ddv440`. Or should we output the same reference twice?

> I'm redoing the greenelab website. For the list of publications we had, maintainers were previously just manually formatting and pasting in the citations. In the new version, we're using manubot cite to automatically get all the meta data like title, author, etc.

Awesome use case for the `manubot cite` utility. Might be time to consider whether to add request caching ability to `manubot cite`.

> However, we want to sort the publications by date

Two thoughts here. The CSL JSON schema is getting updated to allow for a simpler date-field. And we could add a `manubot cite` option that allows sorting (would want to see if there's an easy implementation before committing to this).

Are you using `--format=markdown` or `--format=html` or just the CSL JSON output?

**Comment by vincerubinetti** [comment]
> Are you using `--format=markdown` or `--format=html` or just the CSL JSON output?

Just the json output, because I wasn't sure if pandoc would be required to output yaml, and I don't want anyone editing the repo to have to install anything more than necessary. And markdown/html wouldn't be necessary, because we have to do some custom formatting anyway, which I'm just doing with jekyll.

> A related question, if multiple citations are provided to the same work, should they be consolidated?

I would say do not consolidate them. I'd say preserve order and predictability.

Here's more info on my use case, to help justify that position:

---

We have an `input.yaml` file that looks like this:

```yaml
# preprints
- id: doi:10.1101/2020.05.03.066597
  preprint: true
- id: doi:10.1101/2020.04.14.927251
  github: https://github.com/greenelab/some-repo
- id: arxiv:2003.00898
```

where the goal is, to add a new paper to the Greene Lab publication list, the "user" (the person updating the website) just has to go to this file and add the identifier, and any other meta data they want to include that Manubot wouldn't be able to fetch. Then, the user runs a small python script that loads the yaml, gets the ids, passes it to `manubot cite`, gets the citation data, strips out anything we don't need, cleans it so its easier to be read/sorted by Jekyll, and then outputs an `output.json` file. Then Jekyll can easily read that json data and template out a display for it.

However, because of this issue, that python script has to do extra work to make sure the metadata the passed through correctly from the input to the output, because they are in different orders. I either have to match papers by ids (which might be prone to error due to how many different types of ids there are), or try to pre-sort the input in the same way Manubot would (but I'm not sure if that's alphabetical by "doi:12345" or "12345).

**Comment by vincerubinetti** [comment]
@dhimmel For now, can you tell me whether the id list is sorted alphabetically with or without the id-type prefix (ie `doi:`)? That would be an easy temporary work-around for my use case, for now.

**Comment by dhimmel** [comment]
> can you tell me whether the id list is sorted alphabetically with or without the id-type prefix

It's sorted alphabetically based on the standard_id. The standard_id includes the prefix (in lowercase form). Sometimes the suffix ends up getting changed as part of standardization.

https://github.com/manubot/manubot/blob/47b03e0c202e6a0561e6548ba08a4ad790f56eee/manubot/cite/citations.py#L134

I think I can tackle this within the next couple days if you can wait.

**Comment by vincerubinetti** [comment]
> Sometimes the suffix ends up getting changed as part of standardization.

Ah this is another reason to have it in order. Seems like that would make trying to match input ids to output ids even more fragile.

We can wait, but just in case it takes you longer to get to it, can you tell me if this procedure should theoretically guarantee to sort in the same way manubot does: 1) make list of id strings, prefixes + suffixes 2) make all the strings lowercase 3) use python sort on list.

**Comment by agitter** [comment]
> make all the strings lowercase

This part is a little more complicated.  Manubot is using the "standard_id" for the sorting key.  For many citations, the format of the standard id is the lowercase version if the input string.  For example, `DOI:10.1093/hmg/ddv440` has standard id `doi:10.1093/hmg/ddv440`.  For others, there can be a more complex relationship.  For `https://github.com/manubot/manubot/issues/240` the standard id is `url:https://github.com/manubot/manubot/issues/240`.

This could impact you:
```
$ manubot cite --render https://example.com pmc:PMC4299935
1. Unsupervised feature construction and knowledge extraction from genome-wide assays of breast cancer with denoising autoencoders
Jie Tan, Matthew Ung, Chao Cheng, Casey S Greene
Pacific Symposium on Biocomputing. Pacific Symposium on Biocomputing (2015) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4299935/
PMID: 25592575 · PMCID: PMC4299935
2. Example Domain https://example.com/
```

The URL example is sorted after the PMC example because its standard id starts with `url:` not `https`.

In practice, you'll probably be fine as long as your input list uses the standardized form.  That is, provide URL citations as `url:https://example.com` instead of `https://example.com`.  For any citation, you can quickly check the standard id by running `manubot cite` without `--render`.  The `note` field in the JSON output will show the standard id.

**Comment by vincerubinetti** [comment]
Gotcha.

Given that, I think I'll just wait until Daniel makes the upgrade, to be safe.

**Comment by dhimmel** [comment]
Okay fixed in 3d6d5d4cb79a8674630a5a1ddaafbe04bc08ad80, which I released as part of 0.4.1:

- https://pypi.org/project/manubot/0.4.1/
- https://github.com/manubot/manubot/releases/tag/v0.4.1
